game/armedforces/armedforces.py

```diff
@@ -0,0 +1,82 @@
+from __future__ import annotations
+
+import random
+from typing import TYPE_CHECKING, Iterator, Optional
+from game import db
+from game.data.groups import GroupRole, GroupTask
+from game.armedforces.forcegroup import ForceGroup
+from game.profiling import logged_duration
+
+if TYPE_CHECKING:
+    from game.factions.faction import Faction
+
+
+# TODO More comments and rename
+class ArmedForces:
+    """TODO Description"""
+
+    # All available force groups for a specific Role
+    forces: dict[GroupRole, list[ForceGroup]]
+
+    def __init__(self, faction: Faction):
+        with logged_duration(f"Loading armed forces for {faction.name}"):
+            self._load_forces(faction)
+
+    def add_or_update_force_group(self, new_group: ForceGroup) -> None:
+        """TODO Description"""
+        if new_group.role in self.forces:
+            # Check if a force group with the same units exists
+            for force_group in self.forces[new_group.role]:
+                if (
+                    force_group.units == new_group.units
+                    and force_group.tasks == new_group.tasks
+                ):
+                    # Update existing group if units and tasks are equal
+                    force_group.update_group(new_group)
+                    return
+        # Add a new force group
+        self.add_force_group(new_group)
+
+    def add_force_group(self, force_group: ForceGroup) -> None:
+        """Adds a force group to the forces"""
+        if force_group.role in self.forces:
+            self.forces[force_group.role].append(force_group)
+        else:
+            self.forces[force_group.role] = [force_group]
+
+    def _load_forces(self, faction: Faction) -> None:
+        """Initialize all armed_forces for the given faction"""
+        # This function will create a ForgeGroup for each global Layout and PresetGroup
+        self.forces = {}
+
+        preset_layouts = [
+            layout
+            for preset_group in faction.preset_groups
+            for layout in preset_group.layouts
+        ]
+
+        # Generate Troops for all generic layouts and presets
+        for layout in db.LAYOUTS.layouts:
```

Comment:
This is upside down, isn't it? You're loading `ForceGroup`s for templates, but that's the opposite relationship from the data model. ForceGroups have layouts. Layouts don't have ForceGroups.

I think what you want here is roughly:

```python
for group in ForceGroup.for_faction(faction):
    self.forces[group.role] = group
```

`ForceGroup.for_faction` iterates over all the loaded force groups, filtering by faction compatibility via `faction.can_use_group(group)` (or maybe `group.usable_by_faction(faction)`, idk which will be cleaner but it doesn't matter much).

Force groups would be loaded and cached with something like the following:

```python
for group_decl_file in whatever_glob():
    cls._groups.append(Group.from_yaml(group_decl_file))
```

Which in turn is something like

```python
name = data["name"]
...
for layout_name in data["layouts"]:
    layouts.append(db.LAYOUTS.by_name(layout_name))
return ForceGroup(name, layouts, ...)
```

Reply:
I know what you mean but thats only true for the predefined ForceGroups. The idea here is to get the units which are related to each other (serve the same task together like SHORAD unit and a logistic truck) and the easiest way is to iterate over the `generic` templates and use the related units from them as force group.
If i would reverse it it would basicly achieve the same in the end but just more complicated.

The thing to keep in mind is that acutally ForceGroups are only defined by generic and predefined layouts so why not just iterate over them and create the units from them. Its a way more efficient looping in my opinion